File: strategy/marvin/agent/attacker/penalty.lua
local Base = require "agent/base/behavior"
local Penalty = Class("Agent.Attacker.Penalty", Base)

local World = require "../base/world"

local MoveToStaticBall = require "task/movetostaticball"
local ShootGoal = require "task/shootgoal"
local ShootPenalty = require "task/shootpenalty"


function Penalty:check()
	local mainAttacker = self._inbox.mainAttacker().trainer == self._robot
	local isPenalty = World.RefereeState == "PenaltyOffensivePrepare" or World.RefereeState == "PenaltyOffensive"
	return isPenalty and mainAttacker
end

function Penalty:_updateTask()
	if not self.lookDir then
		self.lookDir = "Right"
		if math.random() < 0.5 then
			self.lookDir = "Left"
		end
	end

	if World.RefereeState == "PenaltyOffensivePrepare" then
		return MoveToStaticBall, {(World.Geometry["OpponentGoal"..self.lookDir] - World.Ball.pos):angle(), 0.08}
	else -- PenaltyOffensive
		return ShootPenalty, {self.lookDir }
	end
end

function Penalty:_stop()
	self.lookDir = nil
end

return Penalty

File: strategy/typescript/glados/agent/attacker/penalty.ts
import * as World from "base/world";

import {Behavior, TaskAssignment} from "glados/agent/base/behavior";
import {MessageType} from "glados/control/messaging";
import {MoveToStaticBall} from "glados/task/attacker/movetostaticball";
import {ShootPenalty} from "glados/task/attacker/shootpenalty";

const G = World.Geometry;

export class Penalty extends Behavior {
	public check (): boolean {
		let mainAttacker = this._messaging.receiveTrainer(MessageType.mainAttacker) === this._robot;
		let isPenalty = World.RefereeState === "PenaltyOffensivePrepare" || World.RefereeState === "PenaltyOffensive";
		return isPenalty && mainAttacker;
	}

	_updateTask (): TaskAssignment<typeof ShootPenalty> | TaskAssignment<typeof MoveToStaticBall> {
		if (World.RefereeState == "PenaltyOffensivePrepare") {
			return [MoveToStaticBall, [(G.OpponentGoal - World.Ball.pos).angle(), 0.08]];
		} else {// PenaltyOffensive
			return [ShootPenalty];
		}
	}
}